ledger/participant-integration-api/src/main/scala/platform/store/backend/common/EventStorageBackendTemplate.scala

```diff
@@ -332,24 +332,54 @@ abstract class EventStorageBackendTemplate(
       fetchSizeHint: Option[Int],
       filterParams: FilterParams,
   )(connection: Connection): Vector[T] = {
-    val parties =
-      filterParams.wildCardParties.iterator
-        .++(filterParams.partiesAndTemplates.iterator.flatMap(_._1.iterator))
-        .map(stringInterning.party.tryInternalize)
-        .flatMap(_.iterator)
-        .toSet
-    SQL"""
+    val internedWildcardParties: Set[Int] = filterParams.wildCardParties.view
+      .flatMap(party => stringInterning.party.tryInternalize(party).toList)
+      .toSet
+
+    val internedPartiesAndTemplates: List[(Set[Int], Set[Int])] =
+      filterParams.partiesAndTemplates.iterator
+        .map { case (parties, templateIds) =>
+          (
+            parties.flatMap(s => stringInterning.party.tryInternalize(s).toList),
+            templateIds.flatMap(s => stringInterning.templateId.tryInternalize(s).toList),
+          )
+        }
+        .filterNot(_._1.isEmpty)
+        .filterNot(_._2.isEmpty)
+        .toList
+
+    val internedAllParties: Set[Int] =
+      internedWildcardParties.concat(internedPartiesAndTemplates.flatMap(_._1))
+
+    if (internedAllParties.isEmpty) {
+      Vector.empty
+    } else {
+      val wildcardPartiesClause = if (internedWildcardParties.isEmpty) {
+        Nil
+      } else {
+        eventStrategy.wildcardPartiesClause(witnessesColumn, internedWildcardParties) :: Nil
+      }
+      val filterPartiesClauses = if (internedPartiesAndTemplates.isEmpty) {
+        Nil
+      } else {
+        eventStrategy.filterPartiesClause(witnessesColumn, internedPartiesAndTemplates)
+      }
+      val witnessesWhereClause =
+        (wildcardPartiesClause ::: filterPartiesClauses).mkComposite("(", " or ", ")")
```

Comment:
In this particular case, your suggestion is nicer. However, I like the direction where all helper methods are dumb (e.g., they expect non-empty inputs) and all special casing is done in the composing method. In your code, you still handle the empty `internedWildcardParties` list and don't push the responsibility of handling it to `eventStrategy.wildcardPartiesClause`.

Good point about the dead code!

Reply:
I am with you on the responsibilities, suggested with that in mind: where we would expect nonempty, we could do .get easily. The ominous .get is exactly what we would need there: communicates clearly to the reader to brace for impact if input col is empty. (this would be also nicer on the error/exception side because it would emit a very precise stacktrace and error as opposed to vague SQL error messages, which are especially vague in case of Oracle)

Reply:
The refactored query methods currently have the following structure:

```scala
val internedStrings = input.map(tryInternalize)
if (internedStrings.isEmpty) {
  Vector.empty
} else {
  SQL"""...""".asVectorOf(parser)
}
```

In 4e10738b64f1866200cc74a01e9eea3b0b9d71dd I have changed `mkComposite` and `arrayIntersectionNonEmptyClause` to throw an exception on empty input. This should never happen with the above structure, but if it does we'll throw a nice exception instead of a vague SQL error.